packages/mds-utils/utils.ts

```diff
@@ -509,30 +509,38 @@ function isInStatesOrEvents(rule: Rule, event: VehicleEvent): boolean {
     return true
   }
 
-  // States that it is possible to transition into with event.event_type
-  const possibleStates: VEHICLE_STATE[] = event.event_types.reduce((acc: VEHICLE_STATE[], event_type) => {
-    return acc.concat(EVENT_STATES_MAP[event_type])
-  }, [])
-  // The last element, assuming the provider didn't make a mistake, should be equivalent
-  // to the state of the event.
-  possibleStates.pop()
-  const result = possibleStates.reduce((acc, state) => {
-    // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    const matchableEvents: any = states[state as VEHICLE_STATE]
+  const { event_types, vehicle_state } = event
+
+  // All event_types except the last (in most cases this will be an empty list)
+  const transientEventTypes = event_types.splice(0, -1)
+
+  /**
+   * State encoded in the event payload (default acc in the reducer) + states that it is
+   * possible to transition into with any transient event_types
+   */
+  const possibleStates: VEHICLE_STATE[] = transientEventTypes.reduce(
+    (acc: VEHICLE_STATE[], event_type) => {
+      return acc.concat(EVENT_STATES_MAP[event_type])
+    },
+    [vehicle_state]
+  )
+
+  return possibleStates.some(state => {
+    const matchableEvents: string[] | undefined = states[state as VEHICLE_STATE]
+
     /* If there's a match between the event_type's transitionable events, and the
-     rule doesn't specify any events, or if there is a match between the rule and the specified
-     events, the rule matches this event. e.g. if the rule says { `available`: [`comms_lost`]} or
-     { `available`: [] }, it would match an event with event_type `comm_lost`.
-    */
+     * rule doesn't specify any events, or if there is a match between the rule and the specified
```

Comment:
Could you update this comment to add something like, "The rule matches this event if an intermediate state and event match, e.g. if the rule says { `reserved`: [] } and there's an event with `event_types` [`trip_end`, `reservation_start`, `trip_start`, `trip_end`], since there's an implication that the vehicle entered the `reserved` state after `reservation_start`, even if the final state of the event is `available`."